Jungun

PARK9898 opened the pull request:
숫자 야구 게임

PARK9898 committed:
숫자 야구 게임

PARK9898 committed:
숫자 야구 게임

0chil commented on src/main/java/baseball/Domain/CompareNumber.java:
```diff
@@ -0,0 +1,29 @@
+
+
+
+	public int Strike(List<Integer> playerNum , List<Integer> comNum){
```
요구사항을 따라주새요.
[코드 컨벤션](https://newwisdom.tistory.com/m/96)이 존재합니다
개발팀의 코드 컨벤션을 따르지 않으면 어떤 문제가 있을까요?

- camelCase
camelCase를 따르며,
클래스명은 CamelCase로,
함수, 변수명은 camelCase로 작성하고 있습니다.

PARK9898 replied:
* 인지는 하고 있었는데 저도 모르게.. 부끄럽네요 고쳤습니다!

0chil replied:
- 개발팀의 코드 컨벤션을 따르지 않으면 어떤 문제가 있을까요?
이 질문도 어떻게 생각하시는지 궁금해요!!

PARK9898 replied:
* 제 생각에는 일단 코드 컨벤션 자체를 따르지 않으면 협엽 자체가 안될것 같아요 이해하기도 어렵고 가독성도 떨어지고요 그리고 코드들을 나중에 다른사람이 유지보수 할 경우에 버그가 발생할 위험도 있겠네요. 개발은 보통 혼자 하는게 아니라 협업하기 때문에 상대방과 서로 이해하기 쉽게 하기 위해 코드 컨벤션을 꼭 지켜주어야겠어요!

0chil commented on src/main/java/baseball/Domain/CompareNumber.java:
```diff
@@ -0,0 +1,29 @@
+
+
+
+	public int Strike(List<Integer> playerNum , List<Integer> comNum){
```
`playerNumber` 대신 `playerNum`으로 축약한 이유,
`playerNumbers` 대신 `playerNum` 으로 단수로 쓴 이유가 있을까요?
다른 사람이 변수명으로 어떤 타입을 넘겨야할지 판단하기가 좀 어렵지 않을까요?

PARK9898 replied:
* 남들이 보기에 확실한 의미 파악이 어렵겠군요! 앞으로는 의미 파악을 쉽게하기 위한 변수명을 사용하겠습니다

PARK9898 replied:
* 땡칠님의 말을 들어보니 무조건 변수명이 짧아야 좋다고 생각했던 제가 싫어지는군요 ㅠㅠ 다른 사람이 봤을 때 이해하기 쉬운 변수명을 채택하도록 노력하겠습니다 (변수명 수정했습니다)

0chil commented on src/main/java/baseball/Domain/CompareNumber.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.List;
+
+public class CompareNumber {
+
+
+
+
+	public int Strike(List<Integer> playerNum , List<Integer> comNum){
```
- 줄바꿈, 공백도 컨벤션입니다
```suggestion

	public int Strike(List<Integer> playerNum, List<Integer> comNum){
```

지금은 어플리케이션이 작으니 상관없지만, 
정언짱의 협업자가 20줄을 띄우고, `,`를 제멋대로 찍는다고 상상해보세요 ㅋㅋ

PARK9898 replied:
* 수정하였습니다!

PARK9898 replied:
* 수정했습니다 앞으로 컨벤션을 잘 지켜 코딩을 하도록 하겠습니다. 나중에 유지보수 할 때도 불편하겠군요 !

0chil commented on src/main/java/baseball/Controller/Game.java:
```diff
@@ -0,0 +1,39 @@
+	private Number playerNumber;
+
+	public Game() {
+		this.computerNumber = Number.ComputerNum();
```
- 생성자를 사용하지 않고 정적 팩토리 메서드를 사용한 이유가 무엇인가요?
- `Game`이 `computerNumber`, `playerNumber`를 필드(상태)로 가질 필요가 있을까요?
  - 게임 객체가 이런 상태를 가지지 않아도, `playGame()`시에 `Number.ComputerNum()`을 호출해서 사용해도 충분하지 않을까요?

PARK9898 replied:
-정적 팩토리 메서드를 사용한 이유는 첫 번째로 Number 객체를 생성할 때 각각의 용도에 맞게 생성되게 하기 위해서이고
다른 이유는 그 용도에 맞는 이름을 사용함으로서 어떠한 객체를 생성했는지 잘 파악할 수 있게 하기 위함입니다.
-필드 상태로는 가질 필요가 없는 것 값습니다 그래서 playGame() 안에서 호출 시 객체를 만들어 사용하도록 코드를 고쳤습니다~

PARK9898 replied:
*정적 팩토리 메서드를 사용한 이유는 첫 번째로 Number 객체를 생성할 때 각각의 용도에 맞게 생성되게 하기 위해서이고
다른 이유는 그 용도에 맞는 이름을 사용함으로서 어떠한 객체를 생성했는지 잘 파악할 수 있게 하기 위함입니다.
*필드 상태로는 가질 필요가 없는 것 값습니다 그래서 playGame() 안에서 호출 시 객체를 만들어 사용하도록 코드를 고쳤습니다~

PARK9898 replied:
- 정적 팩토리 메서드를 사용한 이유는 첫 번째로 Number 객체를 생성할 때 각각의 용도에 맞게 생성되게 하기 위해서이고
다른 이유는 그 용도에 맞는 이름을 사용함으로서 어떠한 객체를 생성했는지 잘 파악할 수 있게 하기 위함입니다.
- 필드 상태로는 가질 필요가 없는 것 값습니다 그래서 playGame() 안에서 호출 시 객체를 만들어 사용하도록 코드를 고쳤습니다~

0chil commented on src/main/java/baseball/Controller/Game.java:
```diff
@@ -0,0 +1,39 @@
+
+	private InputView inputView = new InputView();
+	private OutputView outputView = new OutputView();
+	private CompareNumber compareNumber = new CompareNumber();
```
- 클래스명 `CompareNumber`
compare는 동사입니다. `CompareNumber`는 저에게 '수를 비교한다'라는 행위로 느껴지네요.
클래스라면 '비교 가능한 수' << 이런식으로 표현해주는게 조금 더 적합하지 않을까요?

PARK9898 replied:
* 땡칠님 말을 들어보니 객체 이름에 동사를 쓰는것은 좀 어색한것 같군요 ComparableNumbers로 수정했습니다!

0chil commented on src/main/java/baseball/Controller/Game.java:
```diff
@@ -0,0 +1,39 @@
+		playerNumber = Number.PlayerNum(inputView.inputNumbers());
+
+		outputView.printResult(compareNumber.Strike(computerNumber.getNumbers() ,playerNumber.getNumbers()),
+			compareNumber.Ball(computerNumber.getNumbers() ,playerNumber.getNumbers()));
```
이 부분만 보고는 `Strike()`가 뭔지, `Ball()` 이 뭔지 알 수가 없습니다.
다른 개발자는 무조건 이 메서드를 타고 들어가서 어떤 행위를 하는지 직접 확인해야겠죠.
정언짱의 의도는 '스트라이크 개수를 센다', '볼 개수를 센다' 인 것 같은데, 메서드명에 이 부분을 더 잘 표현해보면 어떨까요?

PARK9898 replied:
* 수정하였습니다!

0chil commented on src/main/java/baseball/Controller/Game.java:
```diff
@@ -0,0 +1,39 @@
+			compareNumber.Ball(computerNumber.getNumbers() ,playerNumber.getNumbers()));
+
+		if(compareNumber.Strike(computerNumber.getNumbers() ,playerNumber.getNumbers()) == 3){
+			outputView.printFinsh();
```
오타가 있네요 🥲

PARK9898 replied:
* 수정하였습니다!

0chil commented on src/main/java/baseball/Domain/CompareNumber.java:
```diff
@@ -0,0 +1,29 @@
+		return strikeCount;
+	}
+
+	public int Ball(List<Integer> playerNum , List<Integer> comNum){
+		int ballCount = 0;
+		for(int i = 0; i < playerNum.size(); i++){
+			if(playerNum.get(i) != comNum.get(i) && comNum.contains(playerNum.get(i))){
+				ballCount++;
+			}
+		}
+		return ballCount;
+	}
```
볼을 세는 '행위', 스트라이크를 세는 '행위'는 
'야구게임의 숫자들'와 연관이 깊어 보이네요.

'숫자들'을 상태로, '스트라이크, 볼을 세는 행위'를 행위로 보고,
한 클래스로 묶으면 한 곳에서 관리하지 편하지 않을까요?

PARK9898 replied:
* 혹시 떙칠님의 말대로라면 스트라이크와 볼을 세는 행위를 하는 메소드를 Numbers class 쪽으로 빼주는게 좋지 않겠냐는 의미인가요??

0chil replied:
@PARK9898 
정확합니다 👍
이건 제 생각일 뿐이니 다르게 생각하시면 다르게 구현하셔도 됩니다.
아래에 정언씨가 구현하신 방법과 이유도 함께 적어주시면 제가 이해가 더 잘 될것 같아요

PARK9898 replied:
제가 일단 Number 클래스에 넣지 않은 이유는 역할을 좀 분리 해주기 위해서입니다. 이러한 로직들이 number class 에 많아지면 책임 또한 많아 지지않을까요?? 제가 생각하기에 스트라이크와 볼을 세는 행위는 하나의 서비스로직 측면?? 이라고 생각해서 분리했습니다. 떙칠님의 의견 듣고싶습니당!

0chil commented on src/main/java/baseball/Domain/Number.java:
```diff
@@ -0,0 +1,49 @@
+import camp.nextstep.edu.missionutils.Randoms;
+
+
+public class Number {
+	private static NumberValidator numberValidator = new NumberValidator();
+	private List<Integer> numbers = new ArrayList<>();
+
+
```
가진건 '숫자들'인데, 클래스명은 '숫자'네요.
다른 개발자가 이 클래스가 여러개의 숫자를 가질거라고 판단하기 어렵지 않을까요?

PARK9898 replied:
* 제가 봐도 Number라는 클래스명을 보면 단수로 보기 쉽겠네요!! 수정하였습니다

0chil commented on src/main/java/baseball/Domain/Number.java:
```diff
@@ -0,0 +1,49 @@
+		this.numbers = numbers;
+	}
+
+	public static Number ComputerNum(){
+
+		List<Integer> ComNumbers = new ArrayList<>();
+		while(ComNumbers.size() < 3){
+			int number = Randoms.pickNumberInRange(1,9);
+			if(!ComNumbers.contains(number)){
+				ComNumbers.add(number);
+			}
+		}
+
+		CheckValidate(ComNumbers);
+
+		return new Number(ComNumbers);
+	}
```
게임에서 랜덤한 숫자를 뽑아서, 생성자를 통해 컴퓨터의 숫자를 생성할 수도 있을 것 같아요.
이 클래스에 이 행위를 담은 정언짱만의 이유가 있을까요?

PARK9898 replied:
* 랜덤으로 숫자를 뽑는 메서드를 따로 분리 할 수도 있었지만 렌덤숫자를 만들어 리스트에 넣는 while 부분은 아주 간단한 로직이기 때문에 따로 빼서 매소드로 따로 만들어 메소드로 만들기 보다는 ComputerNum에 넣어주었습니다 따로 메소드를 분리하는게 더 좋은 방법일까요?? 혹시 그 이유도 알 수 있을까요?

0chil replied:
- 나중에 이 생성자에 수정이나 제거가 필요하다면, 이 생성자에 의존하는 모든 부분을 수정해야 한다. (하위호환성 키워드)

따라서 이런 생성자를 만들 때는 미래까지도 생각해 볼 필요가 있다고 생각합니다.

제 개인적으로는 이 정도로도 충분하다고 생각하는데, '컴퓨터 숫자를 만드는 행위가 숫자의 역할이라고 생각하는 이유'가 듣고싶었습니다.

PARK9898 replied:
음 숫자 자체를 컴퓨터가 만들어야 한다는 편견이 있었던거 같습니다 사실 랜덤 넘버를 만들어서 생성자에 넣어주는게 더 좋은 방법인것 같네요! 수정하였습니다!

0chil commented on src/main/java/baseball/Domain/NumberValidator.java:
```diff
@@ -0,0 +1,34 @@
+	private static final int MAX = 9;
+	private static final int MIN = 1;
+
+	public void checkDuplicate(List<Integer> numbers){
+		Set<Integer> numberSet = new HashSet<>(numbers);
+		if(numberSet.size()!= numbers.size()) {
+			throw new IllegalArgumentException("숫자는 중복되면 안됩니다");
+		}
+	}
+
```
중복 체크에 Set을 잘 활용해주셨네요.
HashSet은 무엇을 기준으로 들어온 객체들의 같고, 다름 (동등성)을 평가할까요?

PARK9898 replied:
Hashset은 객체의 equals 함수를 통해 리턴값이 true이면 논리적으로 같은 객체라고 판단합나다!
이 코드에서는 Integer 객체가 들어가겠군요 Integer객체는 객체 안의 value가 해쉬코드이며 equals() 또한 value의 값으로 객체의 동등성을 판단함으로 중복된 수를 판별할 수 있겠네요!

0chil replied:
1. 우선 'hashCode'로 판단하고, 그 값이 같을 때 equals를 사용하는 것으로 알고있습니다. 

2. 이런 관점에서는 어떤 일이 일어날지 적어봐주시면 좋을 것 같습니다.

PARK9898 replied:
HashSet의 생성자를 호출할 때, 인자로 전달한 numbers 리스트의 요소들을 HashSet에 추가하고 이 과정에서 numbers 리스트의 각 요소의 hashCode() 메서드가 호출되어 각 요소의 해시 코드를 얻습니다 이 때 hashCode()는 Integer객체의 hashCode이기 때문에 Integer객체의 hashCode는 그 객체의 value 값을 리턴합니다 그러면 같은 숫자라면 hashCode가 같다고 판단하고 equals() 메서드를 통해 들어온 객체가 Integer객체의 인스턴스인지 확인하고 Integer객체의 인스턴스이기 때문에 들어온 객체의 value를 int형으로 바꿔주어 value값을 비교하여 같으면 객체가 동등하다고 판단합니다!

PARK9898 replied:
이번을 계기로 좀 더 깊게 공부해야겠다고 생각했어요 ! Integer 객체를 한 번 뜯어보는 시간을 가졌네용!

0chil commented on src/main/java/baseball/Domain/NumberValidator.java:
```diff
@@ -0,0 +1,34 @@
+
+	public void checkRange(List<Integer> numbers){
+		for(Integer a : numbers){
+			if(a < MIN || a > MAX){
```
부등호 방향을 한쪽으로 일치시키면 다른 개발자가 연상하기가 쉬워요.
<img width="462" alt="image" src="https://github.com/PARK9898/java-baseball/assets/39221443/6cab632a-b1dd-49f9-b3e7-6604c0bae59a">

```suggestion
			if(a < MIN || MAX < a){
```

PARK9898 replied:
* 사소하지만 꿀팁이네요! 적용시켰습니다

0chil commented on src/main/java/baseball/View/InputView.java:
```diff
@@ -0,0 +1,42 @@
+		System.out.println("숫자를 입력해주세요 : ");
+		String input = Console.readLine();
+
+		for(int i = 0; i < input.length() ; i++){
+			int number = Character.getNumericValue(input.charAt(i));
+			if(number == -1){
+				throw new IllegalArgumentException("문자가 아닌 숫자를 입력해주세요.");
+			}else{
+				inputNum.add(Character.getNumericValue(input.charAt(i)));
+			}
+		}
+		return inputNum;
```
말씀드렸다시피 공백도 컨벤션입니다.
```suggestion
		for(int i = 0; i < input.length(); i++) {
			int number = Character.getNumericValue(input.charAt(i));
			if(number == -1){
				throw new IllegalArgumentException("문자가 아닌 숫자를 입력해주세요.");
			} else {
				inputNum.add(Character.getNumericValue(input.charAt(i)));
			}
		}
		return inputNum;
```

PARK9898 replied:
* 수정하였습니다!

0chil commented on src/main/java/baseball/View/InputView.java:
```diff
@@ -0,0 +1,42 @@
+public class InputView {
+
+	public List<Integer> inputNumbers(){
+
```
첫 줄을 띄워준 이유가 있나요?

PARK9898 replied:
* 눈에 잘 들어오게 하기 위함이였습니다 안 좋은 방법인가요??

0chil replied:
저는 이미 줄바꿈만으로 잘 이해된다고 생각합니다. (함수 부분)
이건 개인이 선호하는 스타일이 있을 것 같은데, 팀 컨벤션이 있다면 거기에 맞춰주면 서로 더 편할 것 같아요. 👍

0chil commented on src/main/java/baseball/View/InputView.java:
```diff
@@ -0,0 +1,42 @@
+		return inputNum;
+	}
+
+	public boolean StopOrContinue(){
+		System.out.println("게임을 새로 시작하려면 1 종료하려면 2를 눌러주세요");
+		int number = Integer.parseInt(Console.readLine());
+
+		if(number == 1){
+			return true;
+		}
+		if(number == 2){
+			return false;
+		}
+		return StopOrContinue();
+	}
```
`StopOrContinue`라는 이름으로는 true가 stop인지, false가 stop인지 알기가 어려울 것 같아요.

PARK9898 replied:
* 보통 false면 부정이니 stop으로 생각할테니 Continue가 좀 더 적합한 메소드명인것 같네요!

0chil reviewed (changes requested):
안녕하세요 정언짱
코드 잘 봤습니다. 고생 많으셨어요 ㅎㅎ
리뷰 남겼으니 반영하고 다시 만나요!

PARK9898 committed:
Number -> Numbers class 명 변경
PlayerNum -> PlayerNumbers 정적팩토리 메소드명 변경
Checkvalidate -> checkvalidate 함수명 코드 컨벤션 준수

PARK9898 committed:
오타 수정

PARK9898 committed:
코드 컨벤션 준수 & 부등호 방향 수정

PARK9898 committed:
inputNum 변수명 수정

PARK9898 committed:
필드에 있는 객체 메소드로 이동

PARK9898 committed:
Strike, Ball 메소드명 변경

PARK9898 committed:
Strike, Ball 메소드명 변경

PARK9898 commented on src/main/java/baseball/Controller/Game.java:
```diff
@@ -0,0 +1,28 @@
+
+	private InputView inputView = new InputView();
+	private OutputView outputView = new OutputView();
+	private CompareNumber compareNumber = new CompareNumber();
```
* 저는 Game객체 필드 내에서 Inputview와 OutputView를 생성해서 사용했는데 땡칠님 코드를 보면 생성자에 InputView와 OutputView를 매개변수로 받으셨더라구요 혹시 그렇게 한 이유를 알 수 있을까요?

0chil replied:
View의 인스턴스는 어플리케이션 실행과 함께 생성/제거됩니다.
생명주기가 같죠. 
그래서 게임 객체가 인스턴스를 직접 관리하기보다는, 어플리케이션이 인스턴스를 관리해주면 좋겠다고 생각했습니다.

객체 역할(책임) 분리 관점에서는 이런 방향도 좋을 것 같아요.
물론 어디까지나 제 생각입니다. 이 객체에 인스턴스 생성을 맡겨도 전혀 문제없어요!

PARK9898 replied:
확인하였습니다~

PARK9898 committed:
메소드명 수정

0chil commented on src/main/java/baseball/Domain/Numbers.java:
```diff
@@ -0,0 +1,49 @@
+		return new Numbers(ComNumbers);
+	}
+
+	public static Numbers PlayerNumbers(List<Integer> playerNum){
```
정적 팩토리 메서드의 일반적 네이밍 컨벤션을 한번 찾아볼까요?(이펙티브 자바에도 있습니다)
마찬가지로 '메서드' 이기 때문에 명사보다는 동사나 전치사를 사용해주시면 더 이해하기가 쉬울 것 같아요.
특히나 컴퓨터넘버의 경우에는 매번 랜덤으로 새로 생성하니까 그런 의미를 담아주시면 더 좋을 것 같습미다.

PARK9898 replied:
확인하였고 수정하였습니다~

0chil commented on src/main/java/baseball/View/InputView.java:
```diff
@@ -0,0 +1,40 @@
+		return inputNumbers;
+	}
+
+	public boolean Continue(){
```
아직 리뷰가 반영되지 않은 부분이 있네요

PARK9898 replied:
수정하였습니다!

0chil reviewed (changes requested):
정언씨 코드가 훨 깔끔해졌네요. 👍
추가 커멘트를 몇 개 더 작성했는데, 고민해보시고 타당한 것 같다면 반영해주세요~

PARK9898 committed:
메소드명 수정

PARK9898 committed:
메소드명 수정

PARK9898 committed:
메소드 분리

PARK9898 committed:
메소드명 변경

0chil commented on src/main/java/baseball/Domain/NumberValidator.java:
숫자의 검증 책임을 숫자 클래스와 별도로 부여하신 이유가 무엇인가요?

0chil commented on src/main/java/baseball/Domain/Numbers.java:
```diff
@@ -0,0 +1,54 @@
+
+public class Numbers {
+	private static NumberValidator numberValidator = new NumberValidator();
+	private List<Integer> numbers = new ArrayList<>();
```
Final 키워드로 재할당을 막아주어 더 안전하게 만들 수도 있지 않을까요?

0chil commented on src/main/java/baseball/Domain/Numbers.java:
```diff
@@ -0,0 +1,54 @@
+		return new Numbers(playerNum);
+	}
+
+	public List<Integer> getNumbers() {
```
static 메서드 사이에 있으니까 조금 어색하네요🥲
